src/frontend/src/optimizer/plan_node/stream_asof_join.rs

```diff
@@ -0,0 +1,387 @@
+        &self.eq_join_predicate
+    }
+
+    pub(super) fn derive_dist(
+        left: &Distribution,
+        right: &Distribution,
+        logical: &generic::Join<PlanRef>,
+    ) -> Distribution {
+        match (left, right) {
+            (Distribution::Single, Distribution::Single) => Distribution::Single,
+            (Distribution::HashShard(_), Distribution::HashShard(_)) => match logical.join_type {
+                JoinType::Unspecified
+                | JoinType::FullOuter
+                | JoinType::Inner
+                | JoinType::LeftOuter
+                | JoinType::LeftSemi
+                | JoinType::LeftAnti
+                | JoinType::RightSemi
+                | JoinType::RightAnti
+                | JoinType::RightOuter => unreachable!(),
+                JoinType::AsofInner | JoinType::AsofLeftOuter => {
+                    let l2o = logical
+                        .l2i_col_mapping()
+                        .composite(&logical.i2o_col_mapping());
+                    l2o.rewrite_provided_distribution(left)
+                }
+            },
+            (_, _) => unreachable!(
+                "suspicious distribution: left: {:?}, right: {:?}",
+                left, right
+            ),
+        }
+    }
+
+    pub fn derive_dist_key_in_join_key(&self) -> Vec<usize> {
+        let left_dk_indices = self.left().distribution().dist_column_indices().to_vec();
+        let right_dk_indices = self.right().distribution().dist_column_indices().to_vec();
+        let left_jk_indices = self.eq_join_predicate.left_eq_indexes();
+        let right_jk_indices = self.eq_join_predicate.right_eq_indexes();
+
+        assert_eq!(left_jk_indices.len(), right_jk_indices.len());
+
+        let mut dk_indices_in_jk = vec![];
+
+        for (l_dk_idx, r_dk_idx) in left_dk_indices.iter().zip_eq_fast(right_dk_indices.iter()) {
+            for dk_idx_in_jk in left_jk_indices.iter().positions(|idx| idx == l_dk_idx) {
+                if right_jk_indices[dk_idx_in_jk] == *r_dk_idx {
+                    dk_indices_in_jk.push(dk_idx_in_jk);
+                    break;
+                }
+            }
+        }
+
+        assert_eq!(dk_indices_in_jk.len(), left_dk_indices.len());
+        dk_indices_in_jk
+    }
+
+    /// Return stream hash join internal table catalog.
+    pub fn infer_internal_table_catalog<I: StreamPlanRef>(
+        input: I,
+        join_key_indices: Vec<usize>,
+        dk_indices_in_jk: Vec<usize>,
+        inequality_key_idx: usize,
+    ) -> (TableCatalog, Vec<usize>) {
+        let schema = input.schema();
+
+        let internal_table_dist_keys = dk_indices_in_jk
+            .iter()
+            .map(|idx| join_key_indices[*idx])
+            .collect_vec();
+
+        // The pk of AsOf join internal table should be join_key + inequality_key + input_pk.
+        let join_key_len = join_key_indices.len();
+        let mut pk_indices = join_key_indices;
```

Comment:
Do we need to dedup join keys, like what we do for group keys?

Reply:
what do you mean

Reply:
I mean the cases where user write `... JOIN ON a = a AND a = a ...`.

Reply:
Then why we did it for GroupAgg :(

Reply:
Because when users want a query output to identical column, they have to write it twice. If they want to join on a key, there is no reason to write it twice.

Reply:
I got your point now. We deduped the key here https://github.com/singularity-data/risingwave/blob/113ede976b550dcb3b821ad2452c1f8db4dda370/src/frontend/src/utils/condition.rs#L222